feat(peers_acq): shuffle peers before we return.

joshuef opened the pull request:
This shoulf help prevent any one peer from being hammered more than any other in the list

## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 22 Nov 23 09:23 UTC
This pull request adds a new feature to shuffle peers before returning them. This will help prevent any one peer from being hammered more than any other in the list.
<!-- reviewpad:summarize:end -->

grumbach commented on sn_api/src/app/nrs/nrs_map.rs:
```diff
@@ -12,44 +12,44 @@ use log::debug;
 #[derive(Debug, PartialEq, Default, Serialize, Deserialize, Clone)]
 pub struct NrsMap {
-    pub map: BTreeMap<SubName, SafeUrl>,
+    pub map: BTreeMap<Subname, SafeUrl>,
```
The `Subname` type now refers to a full public name, we should probably rename it.

jacderida replied:
Yeah, this is a good suggestion, thanks.

grumbach commented on sn_api/src/app/nrs/nrs_map.rs:
```diff
@@ -12,44 +12,44 @@ use log::debug;
             }
             None => {
-                debug!("NRS: No link found for subname(s): {}", sub_name);
+                debug!("NRS: No link found for subname(s): {}", public_name);
```
And maybe adapt the error messages as well

grumbach commented on sn_api/src/app/nrs/mod.rs:
```diff
@@ -324,53 +304,52 @@ fn convert_multimap_to_nrs_map(multimap: &Multimap) -> Result<NrsMap> {
-fn sanitised_url(name: &str) -> String {
-    format!("safe://{}", name.strip_prefix("safe://").unwrap_or(name))
+fn get_url_from_name(name: &str) -> Result<SafeUrl> {
+    if !name.starts_with("safe://") {
```
Is this check necessairy? A name is not supposed to contain safe:// in the first place rigth?

jacderida replied:
That's done because `SafeUrl::from_url` doesn't work unless the name is prefixed with `safe://`. The calling code uses the URL that's returned.

grumbach replied:
I'm wondering if the `get_url_from_name(name: &str)` `name` here could be confusing, I'm kind of expecting an actual name here as argument (possibly even a `PublicName` type) so without the "safe://", but if you're saying people can call this with a `safe://` scheme+name, then maybe `get_url_from_name` is not the right name? (sorry about all those names)

jacderida replied:
Removed the use of this function. Not quite sure what I was thinking here, to be honest.

grumbach commented on sn_api/src/app/resolver/mod.rs:
````diff
@@ -132,8 +132,8 @@ impl Safe {
     /// ```
-    pub async fn inspect(&self, url: &str) -> Result<Vec<SafeData>> {
-        let safe_url = Safe::parse_url(url)?;
+    pub async fn inspect(&mut self, url: &str) -> Result<Vec<SafeData>> {
````
Why is inspect requiring a `mut` se￼￼lf here?

jacderida replied:
This was a mistake. It was something that was changed in this PR:
https://github.com/maidsafe/safe_network/pull/987/files

I got a conflict with that and I didn't merge it correctly. It should have been removed with my clippy fixes.

grumbach reviewed (commented):
Looking good! A couple of comments, but generally a great improvement!

bochaco commented on sn_cli/src/subcommands/cat.rs:
```diff
@@ -30,17 +30,18 @@ pub struct CatCommands {
-    debug!("Running cat for: {:?}", &url);
+    let link = get_from_arg_or_stdin(cmd.location, None)?;
+    let url = get_target_url(&link)?;
+    debug!("Running cat for: {:?}", &url.to_string());
```
You shouldn't need to call `to_string`, and just use the `{}` for printing it out instead of `{:?}` which is the debug impl.

bochaco commented on sn_cli/src/subcommands/nrs.rs:
```diff
@@ -80,7 +80,7 @@ async fn run_register_subcommand(
 ) -> Result<()> {
     if let Some(ref link) = link {
-        validate_target_link(link)?;
+        get_target_url(link)?;
```
Isn't clippy complaining about not using the returned value?

jacderida replied:
This was left in as a mistake. The function gets called later in the code and the return value is used.

grumbach reviewed (approved):
LGTM

jacderida commented:
bors r+

happybeing commented:
It's there an explanation somewhere if how this differs from the earlier scheme?

It sounds very different, so I'm curious about what is different and the effects it has on applications.

jacderida commented:
> It's there an explanation somewhere if how this differs from the earlier scheme?
> 
> It sounds very different, so I'm curious about what is different and the effects it has on applications.

I'm not sure exactly what you mean by "scheme" in this context, but this just changes how the NRS entries are stored internally; there are no external changes.

happybeing commented:
I mainly understand the change adequately! Previously the public name was used to look up a data structure which contained information for each subname about the 'service' it implemented (e.g. www, or perhaps application defined services), and a pointer (xor address) to the data needed to deliver the service.

On the face of it the service aspect has been dropped, is that the case?

grumbach commented:
Hey @happybeing thanks for that question!

> On the face of it the service aspect has been dropped, is that the case?

No. But we changed the way we access and store data on the NRS map (API side). 

> the public name was used to look up a data structure which contained information for each subname, about the 'service' it implemented

This is still the case, the change in that PR is how we represent that data internally. Previously if one would query "docs.google":

1) we would look for the "google" NRS map container (to simplify, we hash "google" and get a xor address from that)
2) look for the subname "docs" in it
3) and return the SafeUrl associated to it (SafeUrls are typed and associated to a kind of service)

What this PR does is change the way we store data in the NRS map container (step 2) so the steps are the same except for step 2:

2) look for the whole public name (subname+topname) "docs.google" in it

The reason why we decided to store the entire public name in there is because people might query the NrsMap directly using its xor address on the network without knowing it is "google"'s NRS map. What they would get is:
```
"maps" -> Url1
"docs" -> Url2
...
```
They'd have no idea what those subnames refer to, you can't unhash the xor address back to "google" to know it's subnames for google. 
Now, in this PR, since we store the entire public name they will get:
```
"maps.google" -> Url1
"docs.google" -> Url2
...
```

Hope that helped make things clearer?

happybeing commented:
If I've got it, you've potentially reduced the lookups to one in cases (such as www) where the only metadata needed is the type, as that's now part of the URL obtained from the map. Regardless, sounds good. Thank you.

joshuef committed:
feat(peers_acq): shuffle peers before we return.

This shoulf help prevent any one peer from being hammered more than any other in the list
